### PR TITLE
Add Node.js bindings.

### PR DESCRIPTION
**Pull request by abrodersen**
Add Node.js bindings using a fork of Neon.

### CODE REVIEW

File: bindings/node/lib/index.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+function bufferRead(stream) {
+    return new Promise((resolve, reject) => {
+        let buffer = new BufferStream({ 
+            inner: new native.BufferStream() 
+        });
+        stream.on('error', reject);
+        stream.on('end', () => resolve(buffer.inner));
+
+        stream.pipe(buffer);
```
Comment (lurch) [inline bindings/node/lib/index.js:31]: @jhermsmeier This is probably a stupid question, but is / could this be a candidate for using Pipage? ;)

File: bindings/node/lib/index.js [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+                    return Promise.using(disposer, (stream) => bufferWrite(data, stream));
+                }).asCallback(callback);
+            }
+        );
+        this._readValues = Promise.promisify(this._inner.readValues);
+        this._writeValues= Promise.promisify(this._inner.writeValues);
+    }
+
+    loadSchema(json) {
+        let input = JSON.stringify(json);
```
Comment (lurch) [inline bindings/node/lib/index.js]: There seems to be a few places in this file where you're using `let` - I believe (some of) these could be using `const` instead?

File: bindings/node/native/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,23 @@
+[package]
+name = "node-reconfix"
+version = "0.1.0"
+authors = ["Aaron Brodersen <aaron@resin.io>"]
+license = "MIT"
```
Comment (lurch) [inline bindings/node/native/Cargo.toml]: Out of curiosity, how come this is MIT rather than Apache-2.0 ?

File: bindings/node/native/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+build = "build.rs"
+
+[lib]
+name = "node_reconfix"
+crate-type = ["dylib"]
+
+[build-dependencies]
+neon-build = { git = "https://github.com/resin-io-modules/neon.git", branch = "fork" }
```
Comment (lurch) [inline bindings/node/native/Cargo.toml:13]: I wonder if this branch should be given a more meaningful name than just "fork" ?  \*shrug\*

File: bindings/node/native/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+[dependencies]
+neon = { git = "https://github.com/resin-io-modules/neon.git", branch = "fork" }
```
Comment (lurch) [inline bindings/node/native/Cargo.toml:16]: Is it correct that both `neon-build` and `neon` refer to the same git repo?  (it probably is, I'm just checking this isn't a typo)

**Comment by lurch** [comment]
Could this include any unit-tests? Or is that another stupid question? ;)

**Review by majorz — APPROVED** [review]
Lot's of nice work there!

**Comment by majorz** [comment]
Two things that came to my mind during the review: now that we are including JS files, we need to add some automated linting in a future commit. And we need to have automated coverage with a lot of unit tests.

File: bindings/node/lib/index.js (hunk 2 above)
Comment (abrodersen) [reply bindings/node/lib/index.js]: @lurch A habbit formed from writing Rust. :P

File: bindings/node/native/Cargo.toml (hunk 3 above)
Comment (abrodersen) [reply bindings/node/native/Cargo.toml]: @lurch I think Neon generated this and defaults to MIT.

File: bindings/node/native/Cargo.toml (hunk 4 above)
Comment (abrodersen) [reply bindings/node/native/Cargo.toml:13]: @lurch The hope is that this branch will go away soon, and our changes will be mainlined.

File: bindings/node/native/Cargo.toml (hunk 5 above)
Comment (abrodersen) [reply bindings/node/native/Cargo.toml:16]: @lurch Yeah, there are multiple crates in that git repo.

File: bindings/node/native/Cargo.toml (hunk 3 above)
Comment (lurch) [reply bindings/node/native/Cargo.toml]: _Should_ it be changed to Apache2?

**Comment by abrodersen** [comment]
@lurch With regards to unit testing the Node.js bindings, what is the standard approach at Resin? I'm not very knowledgable on the state of unit testing in Node.js.

**Comment by lurch** [comment]
> what is the standard approach at Resin?

I'm not sure - that's probably a better question for @jviotti or @pimterry (or perhaps they'll know who'd be better to ask) ?

**Comment by pimterry** [comment]
[Mocha](https://mochajs.org/) + [chai](http://chaijs.com/) are the standard JS options generally.

Since the vast majority of the code is Rust though, if I were you though I'd focus on automated tests for the rust side, and just add one or two smoke tests in JS to check the basic integration and the tiny bit of JS logic are working correctly. I strongly expect testing Rust in Rust will be a much nicer & more effective experience than trying to poke at it through the JS layer.

The other thing we should look at is TypeScript. Doesn't need to happen for a first release, but the CLI is rapidly turning to TypeScript as we speak, along with most other projects, and before we _deeply_ integrate reconfix into them we'll want to include type definitions (a file that says: 'this module exposes objects A, B, C, with function F that takes X and returns Y'), so we can automatically check we're correctly using it everywhere.

It's fairly easy (the [rindle typings](https://github.com/resin-io/resin-cli/blob/master/typings/rindle.d.ts) in the CLI are a simple example) if you already know the shape of everything. I'm not saying we should do this before merging this PR and getting started, but imo it'd be good to aim for that before we ship any thing with reconfix into prod.

**Pull request merged**